crates/autopilot/src/database/onchain_order_events.rs

```diff
@@ -0,0 +1,727 @@
+        }
+        let mut events_and_quotes = Vec::new();
+        for event in events.iter() {
+            let EthContractEvent { meta, .. } = event;
+            if let Some(meta) = meta {
+                let event_index = meta_to_event_index(meta);
+                if let Some(quote_id) = quote_id_hashmap.get(&event_index) {
+                    events_and_quotes.push((event.clone(), *quote_id));
+                }
+            }
+        }
+        let onchain_order_data = parse_general_onchain_order_placement_data(
+            &*self.quoter,
+            events_and_quotes,
+            self.domain_separator,
+            self.settlement_contract,
+        )
+        .await?;
+        let data_trouple =
+            onchain_order_data
+                .into_iter()
+                .map(|(event_index, onchain_order_placement, order)| {
+                    (
+                        self.custom_onchain_data_parser
+                            .customized_event_data_for_event_index(
+                                &event_index,
+                                &order,
+                                &custom_data_hashmap,
+                                &onchain_order_placement,
+                            ),
+                        (event_index, onchain_order_placement),
+                        order,
+                    )
+                });
+        Ok(data_trouple.unzip_n_vec())
+    }
+}
+
+async fn parse_general_onchain_order_placement_data(
+    quoter: &dyn OrderQuoting,
+    contract_events_and_quotes_zipped: Vec<(EthContractEvent<ContractEvent>, i64)>,
+    domain_separator: DomainSeparator,
+    settlement_contract: H160,
+) -> Result<Vec<(EventIndex, OnchainOrderPlacement, Order)>> {
+    let futures = contract_events_and_quotes_zipped.into_iter().map(
+        |(EthContractEvent { data, meta }, quote_id)| async move {
+            let meta = match meta {
+                Some(meta) => meta,
+                None => return Err(anyhow!("event without metadata")),
+            };
+            let ContractEvent::OrderPlacement(event) = data;
+            let order_data = convert_onchain_order_placement(
+                quoter,
+                &event,
+                quote_id,
+                domain_separator,
+                settlement_contract,
+            )
+            .await?;
+            Ok((meta_to_event_index(&meta), order_data))
+        },
+    );
+    let onchain_order_placement_data = futures::future::join_all(futures).await;
+    onchain_order_placement_data
+        .into_iter()
+        .filter_map(|data| match data {
+            Err(err) => {
+                tracing::debug!("Error while parsing onchain orders: {:}", err);
+                None
+            }
+            Ok((_, None)) => None,
+            Ok((event, Some(order_data))) => Some(Ok((event, order_data.0, order_data.1))),
+        })
+        .collect::<Result<Vec<_>>>()
```

Comment:
Did you think about what happens when there is a temporary error with one of the orders? I don't understand this code well yet but it looks like it would fail the whole function if one order fails. What then? Will it ever try to add the orders again?

Reply:
This is a good comment: I think it was safe, once we had the old event parsing, as it would loop over the event frequently and then retry it automatically. But with Dusan's new event parsing, we are no longer safe.

That said, I think the only likely temporary error is that the quote calculation would fail. But in the ethflow, we will always provide a quote_id; hence this quoting should not be a problem.

Since this is a bigger task, I will implement it in a follow-up PR.